projects/charger/inc/generic_can.h

```diff
@@ -0,0 +1,88 @@
+
+typedef void (*GenericCanRx)(const GenericCanMsg *msg, void *context);
+
+typedef struct GenericCanRxStorage {
+  uint32_t id;
+  GenericCanRx rx_handler;
+  void *context;
+  bool enabled;
+} GenericCanRxStorage;
+
+typedef struct GenericCanInterface {
+  StatusCode (*tx)(const struct GenericCan *can, const GenericCanMsg *msg);
+  // Doesn't support ACKable messages (defaults to enabled).
+  StatusCode (*register_rx)(struct GenericCan *can, GenericCanRx rx_handler, uint32_t id,
+                            void *context);
+  StatusCode (*enable_rx)(struct GenericCan *can, uint32_t id);
+  StatusCode (*disable_rx)(struct GenericCan *can, uint32_t id);
+} GenericCanInterface;
+
+typedef struct GenericCan {
+  GenericCanInterface *interface;
+  GenericCanRxStorage rx_storage[NUM_GENERIC_CAN_RX_HANDLERS];
+} GenericCan;
+
+// Usage:
+//
+// GenericCan uses dynamic dispatch where each struct for an implementation variant of CAN carries
+// with it an implementation of tx, register_rx, enable_rx and disable_rx. To use a given
+// implementation one simply needs to pass the specific type of GenericCan<Network|Hw|Uart> to one
+// of the following functions. Since all three of those types use GenericCan as a base by casting to
+// (GenericCan *). It is possible to derive the interface function pointers without knowing about
+// the rest of the data carried in the struct. This allows an object oriented approach where the
+// remaining variable for a given implementation are "private/protected" meanwhile the interface is
+// consistent for all variants of GenericCan.
+//
+// Example:
+//
+// GenericCanUart can_uart = { 0 };
+// generic_can_uart_init(&can_uart, UART_PORT_1);
+//
+// // Transmit
+// GenericCanMsg my_msg = { 0 };
+//
+// // Populate...
+//
+// generic_can_tx((GenericCan *)&can_uart, &my_msg);
+//
+// The primary advantage of this is that if you build a function that relies on CAN it can accept
+// any GenericCan implementation and be ignorant to its underlying behavior. For example:
+//
+// static uint64_t s_data;
+//
+// void send_internal_data(const GenericCan *can) {
+//   GenericCanMsg my_msg = { .data = &s_data };
+//   generic_can_tx(&can, &my_msg);
+// }
+//
+// In this way the data can be sent via UART, CAN Hw or CAN Network without the caller knowing which
+// specific CAN variant it is dealing with!
+
+// Transmits a GenericCanMsg.
+//
+// Note: if backed by CAN Network |msg| needs to be compatible with CANMessage and its |source_id|
+// will automatically be set to the ID of the broadcaster as determined in |device_id| of a
+// CANSettings struct. Also |type| must be CAN_MSG_TYPE_DATA as this TX mode does not support acks
+// (therefore its CANMessage.msg_id must be in the range [0, 2047].
+StatusCode generic_can_tx(const GenericCan *can, const GenericCanMsg *msg);
+
+// Registers a |rx_handler| (enabled by default) to a |raw_id|.
+StatusCode generic_can_register_rx(GenericCan *can, GenericCanRx rx_handler, uint32_t raw_id,
```

Comment:
Instead of matching purely on raw IDs, maybe a mask/filter combo?

Reply:
Ack. This makes sense. We have very poor documentation of what filtering/masking of CAN means so I had to refer to the datasheet... We should fix that.  

As I understand it this is primarily useful if we only care about matching certain ids (filtering) or a portion of the raw_id (masking) such that we don't process every id. This makes sense for CAN Network especially. 

For the interface it would look something like:
```c
// For all except CAN Network (which already has can_add_filter).
StatusCode geneneric_can_*_init(..., uint32_t bitmask); 
// Only need one mask since in reality there should only be one frame format.

// Then when adding a message id
StatusCode generic_can_register_rx(..., uint32_t id); // not raw_id
// And just process messages matching the id which is what the bitmask
// selects for.
```
Let me know if I have this wrong.

I may opt out of using the mask feature on the charger since there are only two message types anyway but this interface makes more sense.

Reply:
The way it works in `can_hw` is a set of pairs (mask, filter) where `raw_id` is compared: `raw_id & mask == filter`. So if you wanted to only have a single bitmask, you'd actually have a set of pairs where they all have the same mask. I believe your suggested interface is actually incorrect, because it's still comparing the raw ID, just &'d with the bitmask.

I think this abstraction could be helpful in the solar and motor networks, so I think it'd be nice to support mask/filter pairs.

Reply:
Sorry, not sure if I'm following. So my interface is correct iff I switch id back to raw_id?

Reply:
It's semantically correct, but I think including the bitmask in the register RX function is clearer and gives us more options. It's also easier to implement in the case of CAN HW.

Reply:
OK, but providing a bitmask for CAN Network is explicitly invalid then. Would we just ignore that argument?

Reply:
This is part of the reason why I questioned supporting the CAN network in this abstraction layer - it has a lot of differences from CAN HW/UART that I don't think are straightforward to handle. For example, the RX handler has been reimplemented to fit the paradigm and doesn't actually use the CAN network's RX handlers. Similarly, bitmasks could be reimplemented by accessing the CAN HW instance directly, or implementing it in software. At some point, we've implemented a CAN HW layer on top of the CAN network layer.

I'm not really sure that this is a great approach for the CAN network. I fully support this abstraction layer for CAN HW/UART, where I think they're very similar in operation and this layer makes them easy to use. However, the CAN network layer abstracts the actual message format from the user, but this layer requires translating it back into the CAN HW format (raw IDs) for TX, RX, and filtering. I think trying to fit the CAN network into the same interface as CAN HW doesn't work well because they meet different needs.